src/client/metadataApiRetrieve.ts

```diff
@@ -177,28 +182,41 @@ export class MetadataApiRetrieve extends MetadataTransfer<
   }
 
   private async extract(zip: Buffer): Promise<ComponentSet> {
-    const converter = new MetadataConverter(this.options.registry);
-    const { merge, output } = this.options;
-    const outputConfig: ConvertOutputConfig = merge
-      ? {
-          type: 'merge',
-          mergeWith: this.components.getSourceComponents(),
-          defaultDirectory: output,
-        }
-      : {
-          type: 'directory',
-          outputDirectory: output,
-        };
-    const zipComponents = ComponentSet.fromSource({
-      fsPaths: ['.'],
-      registry: this.options.registry,
-      tree: await ZipTreeContainer.create(zip),
-    })
-      .getSourceComponents()
-      .toArray();
-
-    const convertResult = await converter.convert(zipComponents, 'source', outputConfig);
-
-    return new ComponentSet(convertResult.converted, this.options.registry);
+    const components: SourceComponent[] = [];
+    const { merge, output, packageNames, registry } = this.options;
+    const converter = new MetadataConverter(registry);
+    const tree = await ZipTreeContainer.create(zip);
+    // initialize with the specified output and unpackaged metadata
+    const packageAggregator: Package[] = [{ name: output, zipTreeLocation: 'unpackaged' }];
+    packageNames?.forEach((pkg) => {
+      // add the retrieved packages - if they exist
+      packageAggregator.push({ name: pkg, zipTreeLocation: pkg });
+    });
+    const validatedPackages = packageAggregator.filter((pkg) =>
+      tree.isDirectory(pkg.zipTreeLocation)
+    );
```

Comment:
I think we should keep lines 185-188, but I'm not liking the packageAggregator and validated packages code.  We should revert most of that back to how it was originally.  So I think the general flow for this function should be to handle unpackaged metadata, then if there are packageNames defined in the options, handle the packages and aggregate the components.  The `outputConfig` should only be built once since it won't change.

From the previous code, I think we can just change line 193 to: `fsPaths: ['unpackaged']`

Then iterate over any packageNames and build componentSets per package and aggregate it all and convert.

Reply:
Meh...not really liking my comments here.  I was wrong about `outputConfig` - that changes per package retrieved so ignore that.  I think the iteration is fine and afaict it needs to happen that way since the outputConfig, zipComponents, and conversion is computed per package retrieved.  I'm not sure what `validatedPackages` means but seems like that could be moved into the `forEach` above.  There will always be an `unpackaged` dir in the zip due to the manifest.  Maybe change `packageAggregator` array name to `packages`.